File: 1680.concatenation-of-consecutive-binary-numbers/solution.cpp
//
//  C++ / pass : Runtime: 326 ms, faster than 34.24% of C++
//
class Solution {
public:
    int concatenatedBinary(int n) {
        int result = 0;
        int factor = 1;
        int MOD = 1e9+7;
        for (int i = n; i > 0; i--) {
            int j = i;
            while(j > 0) {
                result = (result + ((j & 1) * factor)) % MOD;
                j >>= 1;
                factor = (factor * 2) % MOD;
            }
        }

        return result;
    }
};

File: 1680.concatenation-of-consecutive-binary-numbers/solution.py
#
# Python / TLE
#
# class Solution:
#     def concatenatedBinary(self, n: int) -> int:
#         factor = 1
#         result = 0
#         MOD = 1000000007
#         for i in range(n, 0, -1):
#             while(i > 0):
#                 result = (result + ((i & 1) * factor)) % MOD
#                 i >>= 1
#                 factor = (factor * 2) % MOD
#         return result
#
# print("result = ", Solution().concatenatedBinary(50995))

class Solution:
    def concatenatedBinary(self, n: int) -> int:
        bin_len = 0
        ans = 0
        MOD = 1000000007
        for i in range(1, n + 1):
            if ((i & (i - 1)) == 0):
                bin_len += 1
            ans = ((ans << bin_len) % MOD + i) % MOD
        return ans
